cardano-node/Cardano/Node/CLI.hs

```diff
@@ -93,34 +117,76 @@ data SomeProtocol where
   SomeProtocol :: (RunDemo blk, TraceConstraints blk)
                => Consensus.Protocol blk -> SomeProtocol
 
-fromProtocol :: Protocol -> IO SomeProtocol
-fromProtocol BFT =
+fromProtocol :: CardanoConfiguration -> Protocol -> IO SomeProtocol
+fromProtocol _ BFT =
     case Consensus.runProtocol p of
       Dict -> return $ SomeProtocol p
   where
     p = ProtocolMockBFT defaultSecurityParam
-fromProtocol Praos =
+fromProtocol _ Praos =
     case Consensus.runProtocol p of
       Dict -> return $ SomeProtocol p
   where
     p = ProtocolMockPraos defaultDemoPraosParams
-fromProtocol MockPBFT =
+fromProtocol _ MockPBFT =
     case Consensus.runProtocol p of
       Dict -> return $ SomeProtocol p
   where
     p = ProtocolMockPBFT defaultDemoPBftParams
-fromProtocol RealPBFT =
+fromProtocol CardanoConfiguration{ccCore=ccCore@Core{coStaticKeyMaterial}} RealPBFT = do
+    let Genesis{geSrc, geGenesisHash} = coGenesis ccCore
+        genHash = either (throw . ConfigurationError) id $
+                  decodeAbstractHash geGenesisHash
+        cvtRNM :: RequireNetworkMagic -> RequiresNetworkMagic
+        cvtRNM NoRequireNetworkMagic = RequiresNoMagic
+        cvtRNM RequireNetworkMagic   = RequiresMagic
+        StaticKeyMaterial{skmSigningKeyFile, skmDlgCertFile} = coStaticKeyMaterial
+        kmoDeserialiseDelegateKey = Signing.SigningKey . snd . either (error . show) id . deserialiseFromBytes Signing.fromCBORXPrv
+
+    res <- runExceptT (Genesis.mkConfigFromFile (cvtRNM $ coRequiresNetworkMagic ccCore) geSrc genHash)
+    let geneConfig = case res of
+          Left err -> throw err
+          Right x -> x
+
+    sk  <- kmoDeserialiseDelegateKey <$> LB.readFile skmSigningKeyFile
+    dlg <- either (error . show) id . CanonicalJSON.canonicalDecPre <$> LB.readFile skmDlgCertFile
+
+    let plc = PbftLeaderCredentials sk dlg
+        p   = ProtocolRealPBFT defaultDemoPBftParams geneConfig (pure plc)
+
     case Consensus.runProtocol p of
       Dict -> return $ SomeProtocol p
-  where
-    p = ProtocolRealPBFT defaultDemoPBftParams genesisConfig
-    genesisConfig = Dummy.dummyConfig
+
+-- TODO:  move this to 'cardano-shell'
+instance Exception Genesis.ConfigurationError
 
 -- Node can be run in two modes.
 data ViewMode =
     LiveView    -- Live mode with TUI
   | SimpleView  -- Simple mode, just output text.
 
+{-------------------------------------------------------------------------------
+  Configuration merging
+-------------------------------------------------------------------------------}
+
+-- | Perform merging of layers of configuration, but for now, only in a trivial way,
+--   just for Cardano.Shell.Constants.Types.{Genesis,StaticKeyMaterial}.
+--   We expect this process to become generic at some point.
+mergeConfiguration
+  :: PartialCardanoConfiguration
+  -> Last PartialGenesis
+  -> Last PartialStaticKeyMaterial
+  -> PartialCardanoConfiguration
+mergeConfiguration pcc@PartialCardanoConfiguration{pccCore} pg pskm =
```

Comment:
That is something I'm fixing currently in `cardano-shell`.
This pattern is not that well suited for nested configurations as you can see so we have to keep our configuration at a maximum level of 2 layers if we want it to be somewhat understandable.

Reply:
Or make structures like `PartialGenesis` etc not be `Last` fields, since the leaves of those types are also monoids. Then it should work fine with nested records.

You only need `Last` on a record when that record itself is not a product of monoids. We don't need multiple levels of wrapping. Then merging from the top works very nicely.

Config that sets a single field, and then `<>` at the top level.
```
mainConfig <> mempty { foo = mempty { bar = x } }
```